crypto/key_manager.go

```diff
@@ -82,140 +59,84 @@ func (k PEMKeyManager) HashAlgorithm() crypto.Hash {
-		return err
-	}
-
-	k.serverPrivateKey = key
-	k.signatureAlgorithm = algo
-	signer, err := k.Signer()
-	if err != nil {
-		return err
-	}
-	k.serverPublicKey = signer.Public()
-	return nil
+// Signer returns a signer based on our private key.
+func (k PEMKeyManager) Signer() crypto.Signer {
+	return k.signer
 }
 
-// LoadPublicKey loads a public key from a PEM encoded string.
-func (k *PEMKeyManager) LoadPublicKey(pemEncodedKey string) error {
-	publicBlock, rest := pem.Decode([]byte(pemEncodedKey))
-	if publicBlock == nil {
-		return errors.New("could not decode PEM for public key")
-	}
-	if len(rest) > 0 {
-		return errors.New("extra data found after PEM key decoded")
-	}
-
-	k.rawPublicKey = publicBlock.Bytes
-
-	parsedKey, err := x509.ParsePKIXPublicKey(publicBlock.Bytes)
-	if err != nil {
-		return errors.New("unable to parse public key")
-	}
+// PublicKey returns the public key corresponding to the private key.
+func (k PEMKeyManager) PublicKey() crypto.PublicKey {
+	return k.signer.Public()
+}
 
-	k.serverPublicKey = parsedKey
-	return nil
+// NewFromPrivateKey creates PrivateKeyManager using a private key.
+func NewFromPrivateKey(key crypto.PrivateKey) (PrivateKeyManager, error) {
+	var signer crypto.Signer
+	var sigAlgo sigpb.DigitallySigned_SignatureAlgorithm
+
+	switch key.(type) {
+	case *ecdsa.PrivateKey:
+		signer = key.(crypto.Signer)
+		sigAlgo = sigpb.DigitallySigned_ECDSA
+	case *rsa.PrivateKey:
+		signer = key.(crypto.Signer)
+		sigAlgo = sigpb.DigitallySigned_RSA
+	default:
+		return nil, errors.New("unsupported key type")
+	}
+
+	return &PEMKeyManager{
+		signer:             signer,
+		signatureAlgorithm: sigAlgo,
+	}, nil
 }
 
-// Signer returns a signer based on our private key. Returns an error if no private key
-// has been loaded.
-func (k PEMKeyManager) Signer() (crypto.Signer, error) {
-	if k.serverPrivateKey == nil {
-		return nil, errors.New("private key is not loaded")
+func parsePrivateKey(key []byte) (crypto.PrivateKey, error) {
+	if key, err := x509.ParsePKCS1PrivateKey(key); err == nil {
+		return key, nil
 	}
-
-	// Good old interface{}, this wouldn't be necessary in a proper type system. If it's
-	// even the right thing to do but I couldn't find any useful docs so meh
-	switch k.serverPrivateKey.(type) {
-	case *ecdsa.PrivateKey, *rsa.PrivateKey:
-		return k.serverPrivateKey.(crypto.Signer), nil
+	if key, err := x509.ParsePKCS8PrivateKey(key); err == nil {
+		return key, nil
 	}
-
-	return nil, errors.New("unsupported key type")
-}
-
-// GetPublicKey returns the public key previously loaded or an error if LoadPublicKey has
-// not been previously called successfully.
-func (k PEMKeyManager) GetPublicKey() (crypto.PublicKey, error) {
-	if k.serverPublicKey == nil {
-		return nil, errors.New("called GetPublicKey() but one is not loaded")
+	if key, err := x509.ParseECPrivateKey(key); err == nil {
+		return key, nil
 	}
-
-	return k.serverPublicKey, nil
+	return nil, errors.New("could not parse private key")
```

Comment:
This discards all of the error messages from the parsing functions, which could potentially be helpful.